napari/utils/chunk/_loader.py

```diff
@@ -0,0 +1,480 @@
+            LOGGER.info(
+                "ChunkLoader.clear_pending: %d of %d cleared -> %d remain",
+                num_cleared,
+                num_before,
+                num_after,
+            )
+
+    @staticmethod
+    def _get_request(future: futures.Future) -> Optional[ChunkRequest]:
+        """Return the ChunkRequest for this future.
+
+        Parameters
+        ----------
+        future futures.Future
+            Get the request from this future.
+
+        Returns
+        -------
+        Optional[ChunkRequest]
+            The ChunkRequest or None if the future was cancelled.
+        """
+        try:
+            # Our future has already finished since this is being
+            # called from Chunk_Request._done(), so result() will
+            # not block.
+            return future.result()
+        except futures.CancelledError:
+            LOGGER.info("ChunkLoader._done: cancelled")
+            return None
+
+    def _done(self, future: futures.Future) -> None:
+        """Called when a future finishes with success or was cancelled.
+
+        Parameters
+        ----------
+        future : futures.Future
+            The future that finished or was cancelled.
+
+        Notes
+        -----
+        This method may be called in the worker thread. The documentation
+        very intentionally does not specify which thread the future done
+        callback will be called in, only that it will be called in the same
+        process. So far on MacOS at least it seems to always be called in
+        the worker thread when using a thread pool.
+        """
+        request = self._get_request(future)
+
+        if request is None:
+            return  # Future was cancelled, nothing to do.
+
+        LOGGER.info("ChunkLoader._done: %s", request.key)
+
+        # Now that we are back in this process we can add the perf events.
+        # If we were using threads we could have just added the perf events
+        # normally, but we do it this way just to keep it consistent, and
+        # it should work just as well.
+        request.add_perf_events()
+
+        # Add chunks to the cache in the worker thread. For now it's safe
+        # to do this in the worker. Later we might need to arrange for this
+        # to be done in the GUI thread if cache access becomes more
+        # complicated.
+        self.cache.add_chunks(request)
+
+        # Lookup this Request's LayerInfo.
+        info = self._get_layer_info(request)
+
+        # Resolve the weakref.
+        layer = info.get_layer()
+
+        if layer is None:
+            return  # Ignore chunks since layer was deleted.
+
+        # Record LayerInfo stats.
+        info.load_finished(request, sync=False)
+
+        # Fire event to tell QtChunkReceiver to forward this chunk to its
+        # layer in the GUI thread.
+        self.events.chunk_loaded(layer=layer, request=request)
+
+    def _get_layer_info(self, request: ChunkRequest) -> Optional[LayerInfo]:
```

Comment:
```suggestion
    def _get_layer_info(self, request: ChunkRequest) -> LayerInfo:
```
(since it will raise a KeyError instead of returning `None`)

Reply:
actually... on second glance, do we need this method at all?  since you're not doing anything but grabbing a key from `self.layer_map`... why not just grab it directly? `self.layer_map[request.key.layer_id]`

Reply:
Good catch I got rid of `Optional` and added a **Raises** section the doc string. I like this tiny helper method so we have a nice place to document that it can raise and why it can raise which is now stated as:

```
        # Raises KeyError if not found. This should never happen because we
        # add the layer to the layer_map in ChunkLoader.create_request().
```

We could axe the function but then I think every time we do self.layer_map[layer_id] in the code I'd be tempted to explain what's going there. This way that information is factored out in one place. But that's a judgement call certainly, axing it would not be crazy, but I like it how it is with your changes.